File: visitor/php/Visitor.php
<?php

// Visitor interface
interface ShapeVisitor {
    public function visitCircle(Circle $circle): string;
    public function visitRectangle(Rectangle $rectangle): string;
    public function visitTriangle(Triangle $triangle): string;
}

// Element interface
interface Shape {
    public function accept(ShapeVisitor $visitor): string;
}

// Concrete Elements
class Circle implements Shape {
    public $radius;

    public function __construct(float $radius) {
        $this->radius = $radius;
    }

    public function accept(ShapeVisitor $visitor): string {
        return $visitor->visitCircle($this);
    }
}

class Rectangle implements Shape {
    public $width;
    public $height;

    public function __construct(float $width, float $height) {
        $this->width = $width;
        $this->height = $height;
    }

    public function accept(ShapeVisitor $visitor): string {
        return $visitor->visitRectangle($this);
    }
}

class Triangle implements Shape {
    public $base;
    public $height;

    public function __construct(float $base, float $height) {
        $this->base = $base;
        $this->height = $height;
    }

    public function accept(ShapeVisitor $visitor): string {
        return $visitor->visitTriangle($this);
    }
}

// Concrete Visitors
class AreaCalculator implements ShapeVisitor {
    public function visitCircle(Circle $circle): string {
        $area = pi() * $circle->radius * $circle->radius;
        return sprintf("Circle area: %.2f", $area);
    }

    public function visitRectangle(Rectangle $rectangle): string {
        $area = $rectangle->width * $rectangle->height;
        return sprintf("Rectangle area: %.2f", $area);
    }

    public function visitTriangle(Triangle $triangle): string {
        $area = 0.5 * $triangle->base * $triangle->height;
        return sprintf("Triangle area: %.2f", $area);
    }
}

class DrawVisitor implements ShapeVisitor {
    public function visitCircle(Circle $circle): string {
        return sprintf("Drawing Circle with radius %.2f", $circle->radius);
    }

    public function visitRectangle(Rectangle $rectangle): string {
        return sprintf("Drawing Rectangle %.2f x %.2f", $rectangle->width, $rectangle->height);
    }

    public function visitTriangle(Triangle $triangle): string {
        return sprintf("Drawing Triangle base %.2f, height %.2f", $triangle->base, $triangle->height);
    }
}

// Usage
$shapes = [
    new Circle(5),
    new Rectangle(4, 6),
    new Triangle(3, 4)
];

$areaCalc = new AreaCalculator();
$drawer = new DrawVisitor();

echo "Calculating areas:\n";
foreach ($shapes as $shape) {
    echo $shape->accept($areaCalc) . "\n";
}

echo "\nDrawing shapes:\n";
foreach ($shapes as $shape) {
    echo $shape->accept($drawer) . "\n";
}

File: patterns/visitor/python/Visitor.py
from __future__ import annotations
import math
from abc import ABC, abstractmethod
from typing import List

# Visitor interface
class ShapeVisitor(ABC):
    @abstractmethod
    def visit_circle(self, circle: "Circle") -> str:
        pass

    @abstractmethod
    def visit_rectangle(self, rectangle: "Rectangle") -> str:
        pass

    @abstractmethod
    def visit_triangle(self, triangle: "Triangle") -> str:
        pass

# Element interface
class Shape(ABC):
    @abstractmethod
    def accept(self, visitor: ShapeVisitor) -> str:
        pass

# Concrete Elements
class Circle(Shape):
    def __init__(self, radius: float):
        self.radius = radius

    def accept(self, visitor: ShapeVisitor) -> str:
        return visitor.visit_circle(self)

class Rectangle(Shape):
    def __init__(self, width: float, height: float):
        self.width = width
        self.height = height

    def accept(self, visitor: ShapeVisitor) -> str:
        return visitor.visit_rectangle(self)

class Triangle(Shape):
    def __init__(self, base: float, height: float):
        self.base = base
        self.height = height

    def accept(self, visitor: ShapeVisitor) -> str:
        return visitor.visit_triangle(self)

# Concrete Visitors
class AreaCalculator(ShapeVisitor):
    def visit_circle(self, circle: Circle) -> str:
        area = math.pi * circle.radius ** 2
        return f"Circle area: {area:.2f}"

    def visit_rectangle(self, rectangle: Rectangle) -> str:
        area = rectangle.width * rectangle.height
        return f"Rectangle area: {area:.2f}"

    def visit_triangle(self, triangle: Triangle) -> str:
        area = 0.5 * triangle.base * triangle.height
        return f"Triangle area: {area:.2f}"

class DrawVisitor(ShapeVisitor):
    def visit_circle(self, circle: Circle) -> str:
        return f"Drawing Circle with radius {circle.radius:.2f}"

    def visit_rectangle(self, rectangle: Rectangle) -> str:
        return f"Drawing Rectangle {rectangle.width:.2f} x {rectangle.height:.2f}"

    def visit_triangle(self, triangle: Triangle) -> str:
        return f"Drawing Triangle base {triangle.base:.2f}, height {triangle.height:.2f}"


if __name__ == "__main__":
    shapes: List[Shape] = [
        Circle(5),
        Rectangle(4, 6),
        Triangle(3, 4),
    ]

    area_calc = AreaCalculator()
    drawer = DrawVisitor()

    print("Calculating areas:")
    for shape in shapes:
        print(shape.accept(area_calc))

    print("\nDrawing shapes:")
    for shape in shapes:
        print(shape.accept(drawer))
